File: 127-word-ladder.py
from typing import List
from collections import defaultdict, deque

class Solution:
  def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
    '''
    graph problem with beginword as the first node and endword as the last node
    word list contains the rest of the nodes
    We need to first build the path/edges between each node.
    This is a bidirectional graph where each node can potentially join to other nodes
    if using brute force way to build the edges, we will have O((n^2)*m)
      because we need to go through each node and each character (m) of the word
    Alternative way is to generate a list of keys from each word. Each key will have a single character replaced
    e.g hot -> '*ot', 'h*t', 'ho*'
    we will then use these as the key and the value will be the list of nodes that matches this key.
    This will be the adjacency list i.e the edges we will loop through
    This way, the complexity
    We can then use BFS to loop through each node
    We also need a visited state to mark the node as visited
    
    We also need to check if the endword is in the word list. If not, we return 0
    '''
    if endWord not in wordList:
      return 0

    # Build edges
    adjList = defaultdict(list)
    for word in wordList:
      for j in range(len(word)):
        pattern = word[:j] + '*' + word[j+1:]
        adjList[pattern].append(word)
    visited = set([beginWord])
    queue = deque([beginWord])
    res = 1
    while queue:
      for _ in range(len(queue)):
        word = queue.popleft()
        if word == endWord:
          return res
        # get the edges for each pattern of the word and add those edges into the queue
        for j in range(len(word)):
          pattern = word[:j] + '*' + word[j+1:]
          for neighbor in adjList[pattern]:
            if neighbor not in visited:
              queue.append(neighbor)
              visited.add(neighbor)
      res += 1
    return 0
  
sol = Solution()
beginWord = 'hit'
endWord = 'cog'
wordList = ['hot', 'dot', 'dog', 'lot', 'log', 'cog']
output = sol.ladderLength(beginWord, endWord, wordList)
print(f'beginWord: {beginWord}\nendWord: {endWord}\nwordList: {wordList}\noutput: {output}')

File: 127-word-ladder.java
import java.util.*;

class WordLadderSolution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> visited = new HashSet<>();
        Deque<String> queue = new ArrayDeque<>();
        Map<String, List<String>> hashMap = new HashMap<>();

        boolean containsEndWord = false;
        for (String word : wordList) {
            addWord(word, hashMap, queue, false);
            if (word.equals(endWord)) {
                containsEndWord = true;
            }
        }
        if (!containsEndWord) {
            return 0;
        }

        // Add beginWord
        addWord(beginWord, hashMap, queue, true);

        int counter = 1;
        while (!queue.isEmpty()) {
            counter++;
            String curString;
            for (int i = 0; i < queue.size(); i++) {
                curString = queue.pollFirst();
                visited.add(curString);
                if (curString.equals(endWord)) {
                    return counter;
                }

                for (int j = 0; j < curString.length(); j++) {
                    String key = getKey(j, curString);
                    List<String> neighbours = hashMap.get(key);
                    for (String n : neighbours) {
                        if (!visited.contains(n)) {
                            queue.add(n);
                        }
                    }
                }
            }
        }

        return 0;
    }

    public void addWord(String word, Map<String, List<String>> hashMap, Deque<String> queue, boolean addQueue) {
        for (int i = 0; i < word.length(); i++) {
            String key = getKey(i, word);
            if (!hashMap.containsKey(key)) {
                hashMap.put(key, new ArrayList<>());
            }
            hashMap.get(key).add(word);
            if (addQueue) {
                queue.addAll(hashMap.get(key));
            }
        }
    }

    public String getKey(int i, String word) {
        StringBuilder stringBuilder = new StringBuilder();
        String asterisk = "*";
        stringBuilder.append(word.substring(0, i));
        stringBuilder.append(asterisk);
        stringBuilder.append(word.substring(i + 1));
        return stringBuilder.toString();
    }

    public static void main(String[] args) {
        WordLadderSolution solution = new WordLadderSolution();
        String beginWord = "hit";
        String endWord = "cog";
        List<String> wordList = new ArrayList<>(Arrays.asList("hot", "dot", "dog", "lot", "log", "cog"));
        int result = solution.ladderLength(beginWord, endWord, wordList);
        System.out.println("result: " + result);
    }
}